File: week3/3.1/TheClock/Clock.cs
using System;

namespace TheClock
{
    public class Clock
    {
        private Counter Hours = new();
        private Counter Minutes = new();
        private Counter Seconds = new();

        public string GetTimeString()
        {
            return $"{Hours.Ticks:D2}:{Minutes.Ticks:D2}:{Seconds.Ticks:D2}";
        }

        public void Ticks()
        {
            Seconds.Increment();

            if (Seconds.Ticks >= 60)
            {
                Minutes.Increment();
                Seconds.Reset();
            }

            if (Minutes.Ticks >= 60)
            {
                Hours.Increment();
                Minutes.Reset();
            }

            if (Hours.Ticks >= 24)
            {
                Hours.Reset();
            }
            Thread.Sleep(200);
        }

        public void Reset()
        {
            Hours.Reset();
            Minutes.Reset();
            Seconds.Reset(); 
        }
    }
}

File: week11/11.1/NewClock/Clock.cpp
#include <sstream>
#include <iomanip>
#include "Counter.cpp"

class Clock 
{
    private:
        Counter hours, minutes, seconds;

    public:
        std::string GetTimeString() 
        {
            std::ostringstream oss;  // Use for string formatting
            // Formatted output to ensure two digits with leading zeros if necessary
            oss << std::setw(2) << std::setfill('0') << hours.Ticks() << ":"
                << std::setw(2) << std::setfill('0') << minutes.Ticks() << ":"
                << std::setw(2) << std::setfill('0') << seconds.Ticks();
            return oss.str();  // Convert stream to string and return
        }

        void Ticks() 
        {
            seconds.Increment(); 
            if (seconds.Ticks() >= 60) 
            {
                minutes.Increment();
                seconds.Reset();
            }
            if (minutes.Ticks() >= 60) 
            {
                hours.Increment();
                minutes.Reset();
            }
            if (hours.Ticks() >= 24) 
            {
                hours.Reset();
            }
        }

        void Reset() 
        {
            hours.Reset();
            minutes.Reset();
            seconds.Reset();
        }
};
